### PR TITLE
 Add Loss Scaler classes to the new frontend

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
`amp.loss_Scaler.LossScaler` and `amp.loss_scaler.DynamicLossScaler` classes

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
Depends on #4256

File: orttraining/orttraining/python/training/amp/loss_scaler.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+class LossScaler(object):
+    r"""Base class for implementing custom loss scaler strategies
+
+    Once the scaler is configured, no user intervention is needed to update loss scale during training.
+
+    Note:
+        This class should never be instantiated, but used as an abstract class for custom implementation
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/amp/loss_scaler.py:7]: Can we say "...custom loss scaling strategy."

File: orttraining/orttraining/python/training/amp/loss_scaler.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    def __init__(self):
+        pass
+
+    def reset(self):
+        r"""Resets loss scaler internal state"""
+        raise NotImplementedError
+
+    def update(self, train_step_info):
+        r"""Updates loss scale based on user input and training session info
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/amp/loss_scaler.py:17]: nit: "...scale-based..."

File: orttraining/orttraining/python/training/amp/loss_scaler.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        Args:
+            train_step_info (TrainStepInfo): last step state information
+        """
+        raise NotImplementedError
+
+class DynamicLossScaler(LossScaler):
+    r"""Default implementation for :py:class:`.LossScaler` class used for mixed precision
+
+    This loss scaler works by assuming an initial scale, which is doubled every time a certain number of
+    (stable) training steps are performed without gradients to explode (overflow or reach infinity).
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/amp/loss_scaler.py:28]: nit: "...without exploding gradients..."

File: orttraining/orttraining/python/training/amp/loss_scaler.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        self.loss_scale = self._initial_loss_scale
+        self._stable_steps_count = 0
+
+    def update(self, train_step_info):
+        if train_step_info.all_finite:
+            self._stable_steps_count += 1
+
+            if self._stable_steps_count >= self.up_scale_window:
+                self.loss_scale = min(self.max_loss_scale, self.loss_scale * 2)
+                self._stable_steps_count = 0
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/amp/loss_scaler.py:85]: nit: can `self._stable_steps_count = 0` line be brought out below the if-else block?

File: orttraining/orttraining/python/training/amp/loss_scaler.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        else:
+            self.loss_scale = max(self.min_loss_scale, self.loss_scale / 2)
+            self._stable_steps_count = 0
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/amp/loss_scaler.py:88]: nit: can `self._stable_steps_count = 0` line be brought out below the if-else block?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/amp/loss_scaler.py:88]: I don't think so, as `self._stable_steps_count` shouldn't be zero'ed when `train_step_info.all_finite` is True but `self._stable_steps_count < self.up_scale_window`

File: orttraining/orttraining/python/training/amp/loss_scaler.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/amp/loss_scaler.py:7]: Sure, sometimes line length gets in the way, but your suggestion is still shorter than 128 chars :)

File: orttraining/orttraining/python/training/__init__.py [hunk 6]
```diff
@@ -4,3 +4,4 @@
 #--------------------------------------------------------------------------
 from onnxruntime.capi._pybind_state import TrainingParameters
 from onnxruntime.capi.training.training_session import TrainingSession
+from . import amp
```
Comment (liqunfu) [inline orttraining/orttraining/python/training/__init__.py:7]: nip: what is amp standing for? will it be confused with apex.amp?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/__init__.py:7]: automatic mixed precision and the name matches apex and pytorch amp on purpose, so that users know that onnxrutnime.amp.loss_scaler is similar to apex.amp.loss_scaler, for example
